capsules/src/buzzer_driver.rs

```diff
@@ -0,0 +1,184 @@
+                let interval = (duration_ms as u32) * <A::Frequency>::frequency() / 1000;
+                let tics = self.alarm.now().wrapping_add(interval);
+                self.alarm.set_alarm(tics);
+                ReturnCode::SUCCESS
+            }
+        }
+    }
+
+    fn check_queue(&self) {
+        for appiter in self.apps.iter() {
+            let started_command = appiter.enter(|app, _| {
+                // If this app has a pending command let's use it.
+                app.pending_command.take().map_or(false, |command| {
+                    // Mark this driver as being in use.
+                    self.active_app.set(app.appid());
+                    // Actually make the buzz happen.
+                    self.buzz(command) == ReturnCode::SUCCESS
+                })
+            });
+            if started_command {
+                break;
+            }
+        }
+    }
+}
+
+impl<A: hil::time::Alarm> hil::time::Client for Buzzer<'a, A> {
+    fn fired(&self) {
+        // All we have to do is stop the PWM and check if there are any pending
+        // uses of the buzzer.
+        self.pwm_pin.stop();
+        // Mark the active app as None and see if there is a callback.
+        self.active_app.take().map(|app_id| {
+            let _ = self.apps.enter(app_id, |app, _| {
+                app.callback.map(|mut cb| cb.schedule(0, 0, 0));
+            });
+        });
+
+        // Check if there is anything else to do.
+        self.check_queue();
+    }
+}
+
+/// Provide an interface for userland.
+impl<A: hil::time::Alarm> Driver for Buzzer<'a, A> {
+    /// Setup callbacks.
+    ///
+    /// ### `subscribe_num`
+    ///
+    /// - `0`: Setup a buzz done callback.
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        app_id: AppId,
+    ) -> ReturnCode {
+        self.apps
+            .enter(app_id, |app, _| {
+                match subscribe_num {
+                    0 => app.callback = callback,
+                    _ => return ReturnCode::ENOSUPPORT,
+                }
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into())
+    }
+
+    /// Command interface.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Return SUCCESS if this driver is included on the platform.
+    /// - `1`: Buzz the buzzer. `arg1` is used for the parameters: the lower
+    ///        16 bits determine the frequency in hertz, and the upper 16 bits
+    ///        determine the duration in ms. Note the duration is capped at 5000
+    ///        milliseconds.
+    fn command(&self, command_num: usize, arg1: usize, _: usize, appid: AppId) -> ReturnCode {
+        match command_num {
+            0 => /* This driver exists. */ ReturnCode::SUCCESS,
+
+            1 => {
+                let frequency_hz = arg1 & 0xFFFF;
+                let duration_ms = cmp::min((arg1 >> 16) & 0xFFFF, 5000);
```

Comment:
Might be nice to make `5000` a field of the capsule that the board decides on when initializing the capsule (maybe add a `DEFAULT_MAX_BUZZ` constant just so an indecisive board integrator can go with the value you chose).